### PR TITLE
[APM-662] Improve memory usage / tracking in GraphNodes

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

Improve GraphNode's related memory tracking.
Added extended memory tracking to multiple areas:
- Preparation for Expressions
- ShardMap
- VertexProjections
- EdgeProjections

- [x] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)

#### Related Information

- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1315
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/APM-662

### CODE REVIEW

File: lib/Basics/MemoryTypes/MemoryTypes.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  }
+  [[nodiscard]] size_t operator()(std::string_view txt) const {
+    return std::hash<std::string_view>{}(txt);
+  }
+  [[nodiscard]] size_t operator()(const std::string& txt) const {
+    return std::hash<std::string>{}(txt);
+  }
+
+  std::size_t operator()(MonitoredString const& value) const noexcept {
+    return static_cast<std::size_t>(25);
```
Comment (hkernbach) [inline lib/Basics/MemoryTypes/MemoryTypes.h:49]: WIP

File: arangod/Graph/BaseOptions.h [hunk 2]
```diff
@@ -263,12 +266,14 @@ struct BaseOptions {
...
   // needed for expression evaluation.
   // This entry is required by API, but not actively used here
+  // (currently not monitored as we've not introduced PMR or an alternative yet)
   aql::AqlFunctionsInternalCache _aqlFunctionsInternalCache;
 
   /// This context holds values for Variables/References in AqlNodes
   /// it is read from whenever we need to do a calculation in this class.
   /// e.g. edge.weight > a
   /// Here "a" is read from the expression context.
+  // (currently not monitored as we've not introduced PMR or an alternative yet)
```
Comment (hkernbach) [inline arangod/Graph/BaseOptions.h:276]:
```suggestion
```

File: arangod/Graph/BaseOptions.h [hunk 3]
```diff
@@ -279,6 +284,7 @@ struct BaseOptions {
   /// one LookupInfo.
   /// These list is consulted only if there is no overwrite for a specific depth
   /// so this resambles "ALL ==" parts of filters.
+  // (currently not monitored as we've not introduced PMR or an alternative yet)
```
Comment (hkernbach) [inline arangod/Graph/BaseOptions.h:287]:
```suggestion
```

File: lib/Basics/MemoryTypes/MemoryTypes.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  [[nodiscard]] size_t operator()(std::string_view txt) const {
+    return std::hash<std::string_view>{}(txt);
+  }
+  [[nodiscard]] size_t operator()(const std::string& txt) const {
+    return std::hash<std::string>{}(txt);
+  }
+
+  /*[[nodiscard]] size_t operator()(MonitoredString const& value) const noexcept
+  { return std::hash<std::string_view>{}(std::string_view{value});
+  }*/
```
Comment (hkernbach) [inline lib/Basics/MemoryTypes/MemoryTypes.h:50]:
```suggestion
```

File: arangod/Graph/BaseOptions.cpp [hunk 5]
```diff
@@ -44,6 +46,7 @@
 #include "Graph/TraverserCacheFactory.h"
 #include "Graph/TraverserOptions.h"
 #include "Indexes/Index.h"
+#include "Logger/LogMacros.h"
```
Comment (hkernbach) [inline arangod/Graph/BaseOptions.cpp:49]:
```suggestion
```

File: lib/Basics/MemoryTypes/MemoryTypes.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
+////////////////////////////////////////////////////////////////////////////////
```
Comment (hkernbach) [inline lib/Basics/MemoryTypes/MemoryTypes.h:1]: This file later needs to be merged with Lars' latest changes related to cluster info memory tracking

File: arangod/Graph/BaseOptions.h [hunk 7]
```diff
@@ -230,6 +232,9 @@ struct BaseOptions {
 
   virtual void calculateIndexExpressions(aql::Ast* ast);
 
+  /// @brief return the amount of bytes used
+  size_t getMemoryUsedBytes() const;
+
```
Comment (hkernbach) [inline arangod/Graph/BaseOptions.h:237]:
```suggestion
```

File: arangod/Graph/BaseOptions.cpp [hunk 8]
```diff
@@ -470,10 +484,19 @@ void BaseOptions::serializeVariables(VPackBuilder& builder) const {
...
   _collectionToShard.clear();
   _collectionToShard.reserve(in.size());
   for (auto const& [key, value] : in) {
-    _collectionToShard.emplace(key, std::vector{value});
+    ResourceUsageAllocator<MonitoredCollectionToShardMap> alloc = {
+        _query.resourceMonitor()};
+    auto myVec = MonitoredStringVector{alloc};
+    auto myString = MonitoredString{value, alloc};
+    myVec.emplace_back(myString);
+    _collectionToShard.emplace(key, myVec);
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:499]: Can the values be moved here?
```suggestion
    myVec.emplace_back(myString);
    _collectionToShard.emplace(key, myVec);
```

File: arangod/Graph/BaseOptions.cpp [hunk 9]
```diff
@@ -658,6 +692,29 @@ void BaseOptions::parseShardIndependentFlags(arangodb::velocypack::Slice info) {
...
+    if (!getVertexProjections().empty()) {
+      resourceMonitor().decreaseMemoryUsage(
+          getVertexProjections().size() * sizeof(aql::Projections::Projection));
+    }
+
+    _vertexProjections = Projections::fromVelocyPack(info, "vertexProjections");
+    arangodb::ResourceUsageScope vGuard(
+        resourceMonitor(),
+        getVertexProjections().size() * sizeof(aql::Projections::Projection));
+    vGuard.steal();  // now we are responsible for tracking the memory
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:706]:
```suggestion
      _vertexProjections = Projections::fromVelocyPack(info, "vertexProjections");
      try {
        resourceMonitor().increaseMemoryUsage(getVertexProjections().size() * sizeof(aql::Projections::Projection));
      } catch (...) {
        _vertexProjections.clear();
        throw;
      }
```

File: arangod/Graph/BaseOptions.cpp (hunk 8 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp]: Yes, added. Thanks.

File: arangod/Graph/BaseOptions.cpp (hunk 9 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp]: Thanks, added.

File: arangod/Auth/User.cpp [hunk 10]
```diff
@@ -584,7 +584,7 @@ auth::Level auth::User::collectionAuthLevel(std::string const& dbname,
     if (it != _dbAccess.end()) {
       // Second try to find a specific grant
       CollLevelMap::const_iterator pair =
-          it->second._collectionAccess.find(cname);
+          it->second._collectionAccess.find(std::string{cname});
```
Comment (jsteemann) [inline arangod/Auth/User.cpp:587]: can we save this conversion to string somehow? I guess that the authentication code is called quite often, so it would make sense to make it as quick as possible.

File: arangod/Graph/BaseOptions.cpp [hunk 11]
```diff
@@ -605,11 +619,27 @@ void BaseOptions::isQueryKilledCallback() const {
 }
 
 void BaseOptions::setVertexProjections(Projections projections) {
+  if (!getVertexProjections().empty()) {
+    resourceMonitor().decreaseMemoryUsage(getVertexProjections().size() *
+                                          sizeof(aql::Projections::Projection));
+  }
+  arangodb::ResourceUsageScope guard(
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:626]: if this throws, then we will have an invalid state, because we have already subtracted the memory for the vertex projections, but the vector is not empty. To be on the safe side, the vertex projections vector must be cleared as well.

File: arangod/Graph/BaseOptions.cpp [hunk 12]
```diff
@@ -605,11 +619,27 @@ void BaseOptions::isQueryKilledCallback() const {
...
   _vertexProjections = std::move(projections);
+  guard.steal();  // now we are responsible for tracking the memory
 }
 
 void BaseOptions::setEdgeProjections(Projections projections) {
+  if (!getEdgeProjections().empty()) {
+    resourceMonitor().decreaseMemoryUsage(getEdgeProjections().size() *
+                                          sizeof(aql::Projections::Projection));
+  }
+  arangodb::ResourceUsageScope guard(
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:638]: same here

File: arangod/Graph/BaseOptions.cpp [hunk 13]
```diff
@@ -658,6 +688,39 @@ void BaseOptions::parseShardIndependentFlags(arangodb::velocypack::Slice info) {
...
       DocumentProducingNode::kMaxProjections));
-  _vertexProjections = Projections::fromVelocyPack(info, "vertexProjections");
-  _edgeProjections = Projections::fromVelocyPack(info, "edgeProjections");
-}
+
+  {
+    if (!getVertexProjections().empty()) {
+      resourceMonitor().decreaseMemoryUsage(
+          getVertexProjections().size() * sizeof(aql::Projections::Projection));
+    }
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:696]: must also clear vertex projections here.

File: arangod/Graph/BaseOptions.cpp [hunk 14]
```diff
@@ -658,6 +688,39 @@ void BaseOptions::parseShardIndependentFlags(arangodb::velocypack::Slice info) {
...
+
+    _vertexProjections = Projections::fromVelocyPack(info, "vertexProjections");
+    try {
+      arangodb::ResourceUsageScope vGuard(
+          resourceMonitor(),
+          getVertexProjections().size() * sizeof(aql::Projections::Projection));
+      vGuard.steal();  // now we are responsible for tracking the memory
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:703]: Instead of using a guard and then stealing from it, simply call `increaseMemoryUsage(...)`.

File: arangod/Graph/BaseOptions.cpp [hunk 15]
```diff
@@ -658,6 +688,39 @@ void BaseOptions::parseShardIndependentFlags(arangodb::velocypack::Slice info) {
...
+          getEdgeProjections().size() * sizeof(aql::Projections::Projection));
+    }
+    _edgeProjections = Projections::fromVelocyPack(info, "edgeProjections");
+    try {
+      arangodb::ResourceUsageScope eGuard(
+          resourceMonitor(),
+          getEdgeProjections().size() * sizeof(aql::Projections::Projection));
+      eGuard.steal();  // now we are responsible for tracking the memory
+    } catch (...) {
+      _edgeProjections.clear();
```
Comment (jsteemann) [inline arangod/Graph/BaseOptions.cpp:722]: same as above

File: arangod/StorageEngine/TransactionState.cpp [hunk 16]
```diff
@@ -188,8 +187,8 @@ Result TransactionState::addCollection(DataSourceId cid,
         res.reset(TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION,
                   std::string(TRI_errno_string(
                       TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION)) +
-                      ": " + cname + " [" + AccessMode::typeString(accessType) +
-                      "]");
+                      ": " + std::string{cname} + " [" +
```
Comment (jsteemann) [inline arangod/StorageEngine/TransactionState.cpp:190]: for all these error messages, it would be better to move to `absl::StrCat(...)` to make them more efficient.

File: arangod/Transaction/Methods.cpp [hunk 17]
```diff
@@ -1970,8 +1973,8 @@ DataSourceId transaction::Methods::addCollectionAtRuntime(
           TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION,
           std::string(
               TRI_errno_string(TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION)) +
-              ": " + collectionName + " [" + AccessMode::typeString(type) +
-              "]");
+              ": " + std::string{collectionName} + " [" +
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp:1976]: consider using `absl::StrCat(...)` here.

File: lib/Basics/ResourceUsage.h [hunk 18]
```diff
@@ -146,7 +146,7 @@ class ResourceUsageAllocatorBase : public Allocator {
   using size_type = typename std::allocator_traits<Allocator>::size_type;
   using value_type = typename std::allocator_traits<Allocator>::value_type;
 
-  ResourceUsageAllocatorBase() = delete;
+  ResourceUsageAllocatorBase() = default;
```
Comment (jsteemann) [inline lib/Basics/ResourceUsage.h:149]: is this change still necessary?

File: arangod/Cluster/TraverserEngine.cpp [hunk 19]
```diff
@@ -189,9 +189,10 @@ void BaseEngine::getVertexData(VPackSlice vertex, VPackBuilder& builder,
       return;
     }
     std::string_view vertex = id.substr(pos + 1);
-    for (std::string const& shard : shards->second) {
+    for (auto const& shard : shards->second) {
       Result res = _trx->documentFastPathLocal(
-          shard, vertex, [&](LocalDocumentId const&, VPackSlice doc) {
+          std::string{shard}, vertex,
```
Comment (jsteemann) [inline arangod/Cluster/TraverserEngine.cpp:194]: can we avoid this conversion to string here? I think documentFastPathLocal now accepts a string_view, so we can remove the conversion now.

File: lib/Basics/MemoryTypes/MemoryTypes.h [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+#include <unordered_map>
+
+namespace arangodb {
+typedef std::basic_string<char, std::char_traits<char>,
+                          ResourceUsageAllocator<char, ResourceMonitor>>
+    MonitoredString;
+
+struct hash_monitored_string {
+  using is_transparent = void;
+  [[nodiscard]] size_t operator()(const char* txt) const {
```
Comment (jsteemann) [inline lib/Basics/MemoryTypes/MemoryTypes.h:38]: Our convention is to do it like this:
```suggestion
  [[nodiscard]] size_t operator()(char const* txt) const {
```

File: lib/Basics/MemoryTypes/MemoryTypes.h [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    return std::hash<std::string_view>{}(txt);
+  }
+  [[nodiscard]] size_t operator()(std::string_view txt) const {
+    return std::hash<std::string_view>{}(txt);
+  }
+  [[nodiscard]] size_t operator()(const std::string& txt) const {
```
Comment (jsteemann) [inline lib/Basics/MemoryTypes/MemoryTypes.h:44]: Our convention is to do it like this:
```suggestion
  [[nodiscard]] size_t operator()(std::string const& txt) const {
```

File: arangod/Graph/BaseOptions.cpp (hunk 11 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp]: Added in: 60044d169796e5dd8850ba2454d5f44463b1fc76

File: arangod/Graph/BaseOptions.cpp (hunk 12 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp]: Added in: 60044d169796e5dd8850ba2454d5f44463b1fc76

File: arangod/Graph/BaseOptions.cpp (hunk 13 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp:703]: Added in: 60044d169796e5dd8850ba2454d5f44463b1fc76

File: arangod/Graph/BaseOptions.cpp (hunk 14 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp]: Added in: 60044d169796e5dd8850ba2454d5f44463b1fc76

File: arangod/Graph/BaseOptions.cpp (hunk 15 above)
Comment (hkernbach) [reply arangod/Graph/BaseOptions.cpp]: Added in: 60044d169796e5dd8850ba2454d5f44463b1fc76

File: arangod/Graph/Cache/RefactoredTraverserCache.cpp [hunk 22]
```diff
@@ -307,7 +306,7 @@ bool RefactoredTraverserCache::appendVertex(
               "' as the first line in your AQL");
     }
     for (auto const& shard : it->second) {
-      if (findDocumentInShard(shard)) {
+      if (findDocumentInShard(std::string{shard})) {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Graph/Cache/RefactoredTraverserCache.cpp]: Yes, added in efe727464c8e5114c22a6fc2a3433ece50ac9527

File: arangod/StorageEngine/TransactionState.cpp (hunk 16 above)
Comment (hkernbach) [reply arangod/StorageEngine/TransactionState.cpp]: Added in: efe727464c8e5114c22a6fc2a3433ece50ac9527

File: arangod/Transaction/Methods.cpp (hunk 17 above)
Comment (hkernbach) [reply arangod/Transaction/Methods.cpp]: Added in: efe727464c8e5114c22a6fc2a3433ece50ac9527

File: arangod/Utils/SingleCollectionTransaction.cpp [hunk 23]
```diff
@@ -106,15 +106,15 @@ LogicalCollection* SingleCollectionTransaction::documentCollection() {
...
-    std::string const& name, AccessMode::Type type) {
+    std::string_view name, AccessMode::Type type) {
   TRI_ASSERT(!name.empty());
   if ((name[0] < '0' || name[0] > '9') &&
       name != resolveTrxCollection()->collectionName()) {
     THROW_ARANGO_EXCEPTION_MESSAGE(
         TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION,
         std::string(
             TRI_errno_string(TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION)) +
-            ": " + name);
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Utils/SingleCollectionTransaction.cpp:117]: Added in: efe727464c8e5114c22a6fc2a3433ece50ac9527

File: lib/Basics/ResourceUsage.h (hunk 18 above)
Comment (hkernbach) [reply lib/Basics/ResourceUsage.h:149]: Yes right now it unfortunately is. The compiler complains if not set to default: 

```
/home/hkernbach/Git/ArangoDB/devel-3/tests/Graph/TraverserCacheTest.cpp:58:27: error: call to implicitly-deleted default constructor of 'arangodb::MonitoredCollectionToShardMap' (aka 'unordered_map<basic_string<char, std::char_traits<char>, ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, vector<basic_string<char, std::char_traits<char>, ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, ResourceUsageAllocator<basic_string<char, std::char_traits<char>, ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>, arangodb::hash_monitored_string, arangodb::compare_monitored_string, ResourceUsageAllocator<pair<const basic_string<char, std::char_traits<char>, ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, vector<basic_string<char, std::char_traits<char>, ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, ResourceUsageAllocator<basic_string<char, std::char_traits<char>, ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>>, arangodb::ResourceMonitor>>')
      collectionToShardMap{};  // can be empty, only used in standalone mode
                          ^~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/unordered_map.h:141:7: note: explicitly defaulted function was implicitly deleted here
      unordered_map() = default;
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/unordered_map.h:105:18: note: default constructor of 'unordered_map<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::vector<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceUsageAllocator<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>, arangodb::hash_monitored_string, arangodb::compare_monitored_string, arangodb::ResourceUsageAllocator<std::pair<const std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::vector<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceUsageAllocator<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>>, arangodb::ResourceMonitor>>' is implicitly deleted because field '_M_h' has a deleted default constructor
      _Hashtable _M_h;
                 ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/hashtable.h:531:7: note: explicitly defaulted function was implicitly deleted here
      _Hashtable() = default;
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/hashtable.h:201:7: note: default constructor of '_Hashtable<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::pair<const std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::vector<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceUsageAllocator<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>>, arangodb::ResourceUsageAllocator<std::pair<const std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::vector<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceUsageAllocator<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>>, arangodb::ResourceMonitor>, std::__detail::_Select1st, arangodb::compare_monitored_string, arangodb::hash_monitored_string, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true>>' is implicitly deleted because base class '_Hashtable_enable_default_ctor<arangodb::compare_monitored_string, arangodb::hash_monitored_string, arangodb::ResourceUsageAllocator<std::pair<const std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::vector<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceUsageAllocator<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>>, arangodb::ResourceMonitor>>' (aka '_Enable_default_constructor<__and_<is_default_constructible<arangodb::compare_monitored_string>, is_default_constructible<arangodb::hash_monitored_string>, is_default_constructible<arangodb::ResourceUsageAllocator<std::pair<const std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, std::vector<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceUsageAllocator<std::basic_string<char, std::char_traits<char>, arangodb::ResourceUsageAllocator<char, arangodb::ResourceMonitor>>, arangodb::ResourceMonitor>>>, arangodb::ResourceMonitor>>>{}, __detail::_Hash_node_base>') has a deleted default constructor
      private _Hashtable_enable_default_ctor<_Equal, _Hash, _Alloc>
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/enable_special_members.h:113:15: note: '_Enable_default_constructor' has been explicitly marked deleted here
    constexpr _Enable_default_constructor() noexcept = delete;
              ^
1 error generated.
gmake[2]: *** [tests/Graph/CMakeFiles/arango_tests_graph.dir/build.make:90: tests/Graph/CMakeFiles/arango_tests_graph.dir/TraverserCacheTest.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
```

Do you have a better approach here?

File: arangod/Cluster/TraverserEngine.cpp (hunk 19 above)
Comment (hkernbach) [reply arangod/Cluster/TraverserEngine.cpp]: Good catch! Added in: 9734b57526c551cc0fe071bbc0d89eddae8cf6e6

File: lib/Basics/MemoryTypes/MemoryTypes.h (hunk 21 above)
Comment (hkernbach) [reply lib/Basics/MemoryTypes/MemoryTypes.h]: Added here: 72f4fc70d684953cc3c87e6aaece406cfd7c4b84

File: arangod/Auth/User.cpp (hunk 10 above)
Comment (hkernbach) [reply arangod/Auth/User.cpp:587]: Right now I do not know why, unless we change the data structure of: 
`typedef std::unordered_map<std::string, auth::Level> CollLevelMap;` as well.

File: lib/Basics/ResourceUsage.h (hunk 18 above)
Comment (hkernbach) [reply lib/Basics/ResourceUsage.h:149]: ah - I know how. On it.

Comment (jsteemann) [reply lib/Basics/ResourceUsage.h:149]: We need to provide an allocator from the test, then we can again delete the default constructor.

File: arangod/Auth/User.cpp (hunk 10 above)
Comment (jsteemann) [reply arangod/Auth/User.cpp:587]: What if we give this map a customer hash and eq function? Then we should be able to use heterogeneous lookups.

File: lib/Basics/ResourceUsage.h (hunk 18 above)
Comment (hkernbach) [reply lib/Basics/ResourceUsage.h]: added: 938324ba9ea35224851ff2471843a431aaca15a3

File: arangod/Auth/User.cpp (hunk 10 above)
Comment (hkernbach) [reply arangod/Auth/User.cpp]: Added in abd4ec43b7044cda33327501d291e8cf3f106864

**Review by jsteemann — APPROVED** [review]
LGTM
